src/services/date-utilities.ts

```diff
@@ -0,0 +1,257 @@
+   * @example "2021-01-01 08:00:00 GMT" -> "Jan 1, 2021 at 12:00 pm Pacific time"
+   */
+  static dateToPacificDateTime (date: Date): string {
+    // safety check
+    if (!isDate(date) || isNaN(date.getTime())) return null
+
+    const dateStr = this.dateToPacificDate(date, true)
+    const timeStr = this.dateToPacificTime(date)
+
+    return `${dateStr} at ${timeStr} Pacific time`
+  }
+
+  /**
+   * Converts an API datetime string (in UTC) to an ISO datetime string.
+   * @example 2021-08-05T16:56:50+00:00 -> 2021-08-05T16:56:50Z
+   */
+  static apiToIso (dateTimeString: string): string {
+    if (!dateTimeString) return null // safety check
+
+    // chop off the milliseconds and UTC offset and append "Zulu" timezone abbreviation
+    return dateTimeString.slice(0, 19) + 'Z'
+  }
+
+  /**
+   * Converts an API datetime string (in UTC) to a Date object.
+   * @example 2021-08-05T16:56:50.783101+00:00 -> 2021-08-05T16:56:50Z
+   */
+  static apiToDate (dateTimeString: string): Date {
+    if (!dateTimeString) return null // safety check
+
+    return new Date(this.apiToIso(dateTimeString))
+  }
+
+  /**
+   * Converts an API datetime string (in UTC) to a date and time string (Month Day, Year at HH:MM am/pm
+   * Pacific time).
+   * @example "2021-01-01T00:00:00.000000+00:00" -> "December 31, 2020 at 04:00 pm Pacific time" (PST example)
+   * @example "2021-07-01T00:00:00.000000+00:00" -> "June 30, 2021 at 05:00 pm Pacific time" (PDT example)
+   */
+  static apiToPacificDateTime (dateTimeString: string): string {
+    if (!dateTimeString) return null // safety check
+
+    const date = this.apiToDate(dateTimeString)
+
+    return this.dateToPacificDateTime(date)
+  }
+
+  /**
+   * Converts an API datetime string (in UTC) to a date string (Month Day, Year).
+   * @example "2021-01-01T00:00:00.000000+00:00" -> "December 31, 2020"
+   */
+  static apiToPacificDateLong (dateTimeString: string): string {
+    if (!dateTimeString) return null // safety check
+
+    const date = this.apiToDate(dateTimeString)
+    return this.dateToPacificDate(date, true)
+  }
+
+  /**
+   * Converts a Date object to an API datetime string (in UTC).
+   * @example 2021-08-05T16:56:50Z -> 2021-08-05T16:56:50+00:00
+   */
+  static dateToApi (date: Date): string {
+    // safety check
+    if (!isDate(date) || isNaN(date.getTime())) return null
+
+    // replace "Zulu" timezone abbreviation with UTC offset
+    return date.toISOString().replace('Z', '+00:00')
+  }
+
+  /**
+   * Returns the number of months remaining when given an expiry date string
+   * like this: monthDiffToToday('2023-12-31')
+   * if needed, today's date can be passed as the second parameter
+   */
+  static monthDiffToToday (dateToStr: string, dateFrom = new Date()): number {
+    const dateTo = this.mmmDdYyyyToDate(dateToStr)
+    return dateTo.getMonth() - dateFrom.getMonth() +
+      (12 * (dateTo.getFullYear() - dateFrom.getFullYear()))
+  }
```

Comment:
This is the only new function that's been added to date-utilities.ts

Reply:
Using `new Date()` is dangerous because that returns the browser's date, which might be in a different timezone or wrong/spoofed. Could you make that second parameter required instead?

Reply:
Dangerous? For critical stuff, I would agree.  In this case, the downside risk is the number of months remaining is incorrect which is, I would suggest, of no consequence.  The way it's written a developer can override it if its critical.  I have added a second parameter to `getFormattedExpiryText()` so we can force today's date to a static value for testing etc. I'll push that change shortly.  We should chat about how we handle today's date application wide.  I'm not sure I understand the benefit of pulling today's date from the server.

Reply:
Your current usage of this function may be low risk, but a future developer may reuse it without my caveat above.

Could you default the second parameter to `getCurrentDate` or `getCurrentJsDate` (store getters) instead?

Reply:
The date-utilities service can't reference the store.  This is by design.

I can't see how we can trust the date on the front-end.  Even if we make a call from our server with the current date, the user could change it.   I'm not sure what we're trying to protect by getting the current date and time from the server.  If a user must file a document before midnight, it's the server's date and time that counts.  If the user is late filing, the server returns a business rule error and we gracefully handle that. What am I missing?

Reply:
We have some "future effective" filings, where the user (usually a lawyer/paralegal) specifies an exact date and time to file an IA, for example. We must ensure that when the user says "noon Pacific" that it isn't skewed by a timezone difference or whatever.

I'm thinking that, if you make that second parameter required, we pass the buck to the caller to ensure a valid value, instead of tucking a "dangerous" `new Date()` in this class. Thoughts?

Reply:
In our call yesterday, I think we agreed that there's nothing dangerous about `new Date()` or relying on the browser's date and time.  The [JavaScript documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date) includes no warnings on its use.  We must rely on the API for any current date and time validations.

Reply:
Yes, you're right: using `new Date()` isn't dangerous, but assuming it returns a valid current date/time could cause issues (even though I have yet to come up with a solid example for you).

How about this -- require `dateFrom` above, and in the caller go ahead and use `new Date()` for its values. That satisfies me that the main logic (in the parent) pays attention to what it's doing, and THIS function doesn't make any assumptions.

Reply:
I've actually implemented a method that does the exact same thing in Business Create UI. That's the downside of not having Mixins shared 😟
https://github.com/bcgov/business-create-ui/blob/5e7967885de598886fa6eb3782054e73243d45ac/src/mixins/date-mixin.ts#L260

Reply:
Now, there's going to be inconsistency between Edit UI and Create UI. How should we proceed?

Reply:
> Yes, you're right: using `new Date()` isn't dangerous, but assuming it returns a valid current date/time could cause issues (even though I have yet to come up with a solid example for you).

I think we need to come up with those examples of where `new Date()` could cause problems for our application.  Once we know the problem we're trying to solve, the rest becomes so much easier.  I've designed `monthDiffToToday` so it's nice and clean and the parent doesn't have to provide the current date unless it's a test that requries a static date.  If the browser's current time is using the wrong time zone, the function will off by plus or minus 12 hours when calculating the number of months remaining.  I think we have bigger issues to solve, but once you can cite those examples we'll know for sure.

Reply:
> Now, there's going to be inconsistency between Edit UI and Create UI. How should we proceed?

That's a very good point Karim.  The best way to solve the duplication is to merge the repos.  I can understand that was a good reason to keep the repos separate, now I think there are many more benefits to being able to share each other's good work.

Reply:
I still don't like `new Date()` in a utility function (would prefer higher-level code to make that decision), but you're in this with me so let's merge this and evaluate as we go.

Reply:
> ... The best way to solve the duplication is to merge the repos. I can understand that was a good reason to keep the repos separate, now I think there are many more benefits to being able to share each other's good work.

... Which isn't something we're going to solve in this PR. If the 2 functions are relatively close then let's not worry about them right now. (I think there are other Date Mixin differences anyway.)